### ISSUE TITLE
Incorrect errors when adding `this` parameter to functions

### ISSUE THREAD

**Issue by olee** [palantir/tslint#1393]
### Bug Report

- __TSLint version__: (unknown - tslint used by Atom plugin `linter-tslint`)
- __TypeScript version__: latest
- __Running TSLint via__: Atom

#### TypeScript code being linted

```ts
function cloneEntity(this: IEntity) {
    var result = Object.create(Object.getPrototypeOf(this));
    Object.keys(this).filter(key => !key.startsWith('$') && filteredEntityProperties.indexOf(key) < 0).forEach((key) => {
        result[key] = angular.copy(this[key]);
    });
    return result;
}
```

#### Actual behavior

The following errors are show (in Atom)
````
no-unused-expression - expected an assignment or function callat line 22 col 22
no-unused-expression - expected an assignment or function callat line 22 col 28
semicolon - missing semicolonat line 22 col 26
semicolon - missing semicolonat line 22 col 35
````

#### Expected behavior

No error shown

**Comment by adidahiya** [comment]
Can you please dig in and find the tslint & typescript version actually in use? By "latest" do you mean v2.0? I think this has to do with some new syntax introduced recently into the language.

**Comment by olee** [comment]
The definition of a `this` type in functions was added in [this PR](https://github.com/Microsoft/TypeScript/pull/6739).
I think tslint probably just doesn't support it yet.

**Comment by olee** [comment]
Do you still need more information?
I don't really know what else I could provide you.
It would be really nice if tslint could handle those cases because right now I get warnings everywhere...

**Comment by adidahiya** [comment]
@olee have you tryied using `tslint@next`? See this bit of the readme about using latest versions of the compiler: https://github.com/palantir/tslint#peer-dependencies

**Comment by olee** [comment]
How would I go about using tslint@next in Atom-typescript so I can try it?
Does someone know that?

**Comment by JKillian** [comment]
You could try `npm install -g tslint@next` and try running it on your code via the CLI. If you can duplicate the bug via the CLI, it's definitely a TSLint bug

**Comment by olee** [comment]
Indeed, using tslint together with typescript@next seems to work properly with this test code:
````
function formatShortDate(this: Date) {
    return `${this.getFullYear()}-${this.getMonth()}-${this.getDate()}`;
}

Date.prototype.formatShortDate = formatShortDate;

console.log(new Date().formatShortDate());
````

Now I just need to figure out how to get atom-tslint to use typescript@next (or tslint@next) as well...

**Comment by JKillian** [comment]
Glad it's not a TSLInt bug! Going to close this issue, but feel free to continue discussion here if needed
